src/addon/addonmanager.js

```diff
@@ -6,134 +6,281 @@
+                        self.project = project;
 
-        startParams = {
-            projectId: projectId,
-            branchName: branchName,
-            logger: logger.fork(addOnName)
-        };
+                        if (rights.write === false) {
+                            logger.warn('AddOnManager for project [' + projectId +
+                                '] initialized without write access.');
+                        }
 
-        addOn.start(startParams, function (err) {
-            if (err) {
-                deferred.reject(err instanceof Error ? err : new Error(err));
-            } else {
-                runningAddOns[addOnName] = addOn;
-                deferred.resolve();
-            }
-        });
+                        self.initialized = true;
+                        initDeferred.resolve();
+                    });
+                } else {
+                    logger.error(new Error('Connection problems' + networkStatus));
+                    initDeferred.reject(new Error('Problems connecting to the webgme server, network status: ' +
+                        networkStatus));
+                }
+            });
+        }
 
-        return deferred.promise.nodeify(callback);
+        return initDeferred.promise.nodeify(callback);
     };
 
-    this.queryAddOn = function (addOnName, parameters, callback) {
-        var deferred = Q.defer(),
-            addOn;
+    this.monitorBranch = function (webGMESessionId, branchName, callback) {
+        var monitor = self.branchMonitors[branchName],
+            deferred = Q.defer();
 
-        if (!addOnName) {
-            //TODO: This assumes a single running addOn.
-            addOnName = Object.keys(runningAddOns)[0];
-            logger.debug('No addOnName given for query picked one randomly', addOnName);
-        }
+        function startNewMonitor(initCnt) {
+            monitor = {
+                connectionCnt: initCnt,
+                stopTimeout: null,
+                instance: new BranchMonitor(webGMESessionId, self.storage, self.project, branchName, logger, gmeConfig)
+            };
 
-        addOn = runningAddOns[addOnName];
+            self.branchMonitors[branchName] = monitor;
+            logger.debug('monitorBranch [' + branchName + '] - starting new monitor');
+            monitor.instance.start()
+                .then(deferred.resolve)
+                .catch(deferred.reject);
+        }
 
-        if (!addOn) {
-            deferred.reject(new Error('The addOn is not running'));
-        } else {
-            addOn.query(parameters, function (err, message) {
-                if (err) {
-                    deferred.reject(err instanceof Error ? err : new Error(err));
+        if (monitor) {
+            monitor.connectionCnt += 1;
+            logger.debug('monitorBranch [' + branchName + '] - connection counter [prev, now]',
+                monitor.connectionCnt - 1, monitor.connectionCnt);
```

Comment:
is it worth printing out n, n+1 all the time? (I mean if that is the only option, it seems a bit redundant)